### PR TITLE
Refine the logic for scanning Julia stacks.

### PR DESCRIPTION
**Pull request by rbehrends**
This fixes how we handle scanning the root task of the main thread. It also improves the comments that explain stack scanning logic.

### CODE REVIEW

File: src/julia_gc.c [hunk 1]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
 // from which freed objects are then marked. Hence, we add additional checks
 // when traversing GAP master pointer and bag objects that this happens
 // only for live objects.
+//
+// All stacks in Julia are associated with tasks and we can use
+// jl_task_stack_buffer() to retrieve the buffer information for that stack.
+// That said, in a couple of cases these need adjustments.
+//
+// 1. The stack buffer of the root task of the main thread, when started from
+//    GAP can extend past the point where Julia believes it ends. Therefore,
```
Comment (fingolfin) [inline src/julia_gc.c:397]: Shouldn't this be:
```suggestion
//    GAP can extend past the point where Julia believes it starts. Therefore,
```
or
```suggestion
//    GAP can extend past the point where Julia believes its origin is. Therefore,
```
or so?

File: src/julia_gc.c [hunk 2]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+//    for that stack, we use GapBottomStack instead.
+// 2. If it is the current task of the current thread, we know where exactly
+//    the lowest valid address of the stack is and do not need to scan the
```
Comment (fingolfin) [inline src/julia_gc.c:400]:
```suggestion
//    the top (i.e.,  on all currently popular architectures: lowest valid address) of the stack is and do not need to scan the
```

File: src/julia_gc.c [hunk 3]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+// In addition, stack buffers can include guard pages, which trigger a
+// segmentation fault when accessed. We intercept such segmentation faults
+// and terminate stack scanning early for those.
+//
+// In GAP terminology, we use "bottom" to describe the end of the stack
+// where it starts and "top" to describe the current stack pointer. This
+// means that the "bottom" of the stack is the highest address on most
+// modern architectures and "top" is the lowest. When we treat the stack
+// as an abstract buffer, "start" and "end" refer to its lowest and
+// highest address, respectively.
```
Comment (fingolfin) [inline src/julia_gc.c:412]: That's not just "GAP terminology". I found this use of top and bottom in virtually all CS references I checked. Also, I'd bring these definitions earlier, then the text before here could use them instead of talking about origin/start/end/lowest address etc.

Also, I think "the end ... where it starts" is a really confusing choice of words. Perhaps: This:

```suggestion
// We use "bottom" to refer to the origin of the stack, and "top" to describe
// the current stack pointer. Confusingly, on most contemporary architectures,
// the stack grows "downwards", which means that the "bottom" of the stack
// is the highest address and "top" is the lowest. When we treat the stack
// as an abstract buffer, "start" and "end" refer to its lowest and
// highest address, respectively.
```
Also, perhaps instead of "start" and "end" it would then be better to talk about "low" and "high" ?

File: src/julia_gc.c [hunk 4]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+//
+//   +------------------------------------------------+
+//   | guard |  unused area            | active stack |
+//   | pages |                         | frames       |
```
Comment (fingolfin) [inline src/julia_gc.c:416]: How about adding in the growth direction?
```suggestion
//   | pages |     <--- growth ---     | frames       |
```

File: src/julia_gc.c [hunk 5]
```diff
@@ -653,7 +685,16 @@ static void GapTaskScanner(jl_task_t * task, int root_task)
             stack = stack + size - task->copy_stack;
             size = task->copy_stack;
         }
-        ScanTaskStack(rescan, task, stack, stack + size);
+        char * stackend = stack + size;
+        if (task == RootTaskOfMainThread) {
+            stackend = (char *)GapStackBottom;
+        }
+        // Unlike the stack of the current task that we can in
```
Comment (fingolfin) [inline src/julia_gc.c:692]:
```suggestion
        // Unlike the stack of the current task that we scan in
```

File: src/julia_gc.c [hunk 6]
```diff
@@ -765,6 +806,22 @@ void InitBags(UInt initial_size, Bag * stack_bottom, UInt stack_align)
...
     jl_init();
 
+    JuliaTLS = jl_get_ptls_states();
+    // These callbacks potentially require access to the Julia
+    // TLS and thus need to be installed after initialization.
+    jl_gc_set_cb_root_scanner(GapRootScanner, 1);
+    jl_gc_set_cb_task_scanner(GapTaskScanner, 1);
+    jl_gc_set_cb_pre_gc(PreGCHook, 1);
+    jl_gc_set_cb_post_gc(PostGCHook, 1);
+    // jl_gc_enable(0); /// DEBUGGING
```
Comment (fingolfin) [inline src/julia_gc.c:816]: Why was this moved up?

File: src/julia_gc.c [hunk 7]
```diff
@@ -765,6 +806,22 @@ void InitBags(UInt initial_size, Bag * stack_bottom, UInt stack_align)
...
+
+    // If we are embedding Julia in GAP, remember the root task
+    // of the main thread. The extent of the stack buffer of that
+    // task is calculated a bit differently than for other tasks.
+
```
Comment (fingolfin) [inline src/julia_gc.c:821]:
```suggestion
```

File: src/julia_gc.c (hunk 6 above)
Comment (rbehrends) [reply src/julia_gc.c:816]: These calls were originally directly after `jl_init()` (see comment). In between then and now, code was reorganized that moved them down and can in theory (though not currently in practice) trigger a GC for which they would be needed. This change is putting them back to where they were before.

File: src/julia_gc.c (hunk 1 above)
Comment (rbehrends) [reply src/julia_gc.c:397]: I like the second option best.

File: src/julia_gc.c (hunk 2 above)
Comment (rbehrends) [reply src/julia_gc.c:400]: This would unfortunately be inaccurate. The top of the stack is the current stack pointer, not the beginning of the stack buffer (which is an implementation detail).

File: src/julia_gc.c (hunk 3 above)
Comment (rbehrends) [reply src/julia_gc.c:412]: It's conventional terminology when dealing with stacks as abstract data types; when dealing with hardware stacks, terminology is unfortunately inconsistent. Julia in particular uses the opposite terminology for its task stacks ([example](https://github.com/JuliaLang/julia/blob/6cd329c371c1db3d9876bc337e82e274e50420e8/src/task.c#L1058)). This is why I tried to distinguish between Julia and GAP terminology.

However, I think I still like your wording better, as it's overall clearer and more to the point, so I'm good with the change.

File: src/julia_gc.c (hunk 4 above)
Comment (rbehrends) [reply src/julia_gc.c:416]: Good idea.

File: src/julia_gc.c (hunk 5 above)
Comment (rbehrends) [reply src/julia_gc.c:692]: Oops.

File: src/julia_gc.c (hunk 7 above)
Comment (rbehrends) [reply src/julia_gc.c:821]: Done.

File: src/julia_gc.c [hunk 8]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+//   +------------------------------------------------+
+//   | guard |    unused area          | active stack |
+//   | pages |  <--- growth ---        | frames       |
+//   +------------------------------------------------+
+//   ^                                 ^              ^
+//   |                                 |              |
+// start                              top         bottom/end
+//
+// All stacks in Julia are associated with tasks and we can use
+// jl_task_stack_buffer() to retrieve the buffer information for that stack.
```
Comment (fingolfin) [inline src/julia_gc.c:408]: With the definitions above, we could make this more precise, right?
```suggestion
// jl_task_stack_buffer() to retrieve the start and end pointer for that stack.
```

File: src/julia_gc.c [hunk 9]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+// That said, in a couple of cases these need adjustments.
+//
+// 1. The stack buffer of the root task of the main thread, when started
+//    from GAP can extend past the point where Julia believes its bottom is.
+//    Therefore, for that stack, we use GapBottomStack instead.
+// 2. If it is the current task of the current thread, we know where exactly
```
Comment (fingolfin) [inline src/julia_gc.c:414]:
```suggestion
// 2. For the current task of the current thread, we know where exactly
```

File: src/julia_gc.c [hunk 10]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+//   | guard |    unused area          | active stack |
+//   | pages |  <--- growth ---        | frames       |
+//   +------------------------------------------------+
+//   ^                                 ^              ^
+//   |                                 |              |
+// start                              top         bottom/end
+//
+// All stacks in Julia are associated with tasks and we can use
+// jl_task_stack_buffer() to retrieve the buffer information for that stack.
+// That said, in a couple of cases these need adjustments.
```
Comment (fingolfin) [inline src/julia_gc.c:409]:
```suggestion
// That said, in a couple of cases we make adjustments:
```
Note that you give one case where adjustments are needed, and one were it is an optimization.

File: src/julia_gc.c [hunk 11]
```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
...
+//
+// 1. The stack buffer of the root task of the main thread, when started
+//    from GAP can extend past the point where Julia believes its bottom is.
+//    Therefore, for that stack, we use GapBottomStack instead.
+// 2. If it is the current task of the current thread, we know where exactly
+//    the top is and do not need to scan the entire stack buffer.
+//
+// In addition, stack buffers can include guard pages, which trigger a
+// segmentation fault when accessed. We intercept such segmentation faults
+// and terminate stack scanning early for those.
```
Comment (fingolfin) [inline src/julia_gc.c:419]: From this, it does not become clear (a) where the guard pages are (your sketch above of course shows this, but perhaps mention it again), but more importantly (b) it is unclear why Julia would point <start> beyond the guard page... And it's not clear to me if it really *does* that? I thought the problem was only for root tasks, as their stack is not allocated by Julia? Or what? Sorry, I am still confused about this :-/

File: src/julia_gc.c [hunk 12]
```diff
@@ -491,7 +522,7 @@ static void SafeScanTaskStack(PtrArray * stack, void * start, void * end)
         // longjmps to JuliaTLS->safe_restore; we use this
         // mechamism to abort stack scanning when a protected
         // page is hit. For this to work, we must scan the stack
-        // from top to bottom, so we see any guard pages last.
+        // from bottom to top, so we see any guard pages last.
```
Comment (fingolfin) [inline src/julia_gc.c:525]: this doesn't fit the terminology above anymore
```suggestion
        // from start to end (as defined above), so we see any guard pages last.
```

File: src/julia_gc.c [hunk 13]
```diff
@@ -653,7 +685,16 @@ static void GapTaskScanner(jl_task_t * task, int root_task)
             stack = stack + size - task->copy_stack;
             size = task->copy_stack;
         }
-        ScanTaskStack(rescan, task, stack, stack + size);
+        char * stackend = stack + size;
+        if (task == RootTaskOfMainThread) {
+            stackend = (char *)GapStackBottom;
+        }
```
Comment (fingolfin) [inline src/julia_gc.c:691]:
```suggestion
        }

```

Comment (rbehrends) [reply src/julia_gc.c:691]: Done.

File: src/julia_gc.c (hunk 12 above)
Comment (rbehrends) [reply src/julia_gc.c:525]: Umm, I don't see it or maybe I'm misunderstanding something? The guard pages are at the start of the buffer. I agree that "to top" is misleading and should be "towards the start of the stack buffer" or something like that. But the direction is high -> low address?

File: src/julia_gc.c (hunk 11 above)
Comment (rbehrends) [reply src/julia_gc.c:419]: The Julia stack buffer information includes the guard pages. The information that Julia internally maintains is the start and end of the stack buffer, including guard pages. For some stacks, Julia may not even know where the guard pages end. This can in principle (with `COPY_STACKS`) be any task; in addition, it may change in future implementations.

I will try and spell out more clearly what we're doing here.

File: src/julia_gc.c (hunk 12 above)
Comment (fingolfin) [reply src/julia_gc.c:525]: The thing is that in here, we don't necessary scan till the top (as we don't know where it is). As to whether we scan from "end to start" or "start to end" , I forgot that we scan "downwards"; so we could say "from end downwards to the start". But it would indeed nice to say "from bottom to top"; alas, again, I am concerned that "top" is not accurate here in general?

We could also think about calling `start` something else, like `max_top` or `absolute_top` or so?

File: src/julia_gc.c (hunk 10 above)
Comment (rbehrends) [reply src/julia_gc.c:409]: Hmm. I was using adjustments as a general term (synonym for changes) that included optimizations. In any event, I've altered the comment to include the new phrasing.

File: src/julia_gc.c (hunk 9 above)
Comment (rbehrends) [reply src/julia_gc.c:414]: Done.

File: src/julia_gc.c (hunk 8 above)
Comment (rbehrends) [reply src/julia_gc.c:408]: It's actually start and size that we get, so how about:

```
// All stacks in Julia are associated with tasks and we can use
// jl_task_stack_buffer() to retrieve the buffer information (start and size)
// for that stack.
```

File: src/julia_gc.c (hunk 12 above)
Comment (rbehrends) [reply src/julia_gc.c:525]: So, I've rephrased comments to refer to the start/end of the *stack buffer* that contains the stack, and the top/bottom of the *stack* and also did some rewording to describe the actual process.

File: src/julia_gc.c (hunk 8 above)
Comment (fingolfin) [reply src/julia_gc.c]: fine by me

File: src/julia_gc.c [hunk 14]
```diff
@@ -647,13 +684,25 @@ static void GapTaskScanner(jl_task_t * task, int root_task)
             rescan = 0;
     }
     if (stack) {
-        if (task->copy_stack) {
+        // task->copy_stack, if not zero or one, is an upper bound for the
+        // size of the stack.
+        if (task->copy_stack > 1) {
```
Comment (fingolfin) [inline src/julia_gc.c:689]: I am not sure this is quite right now -- the value `task->copy_stack == 1` should probably still be treated separately? I think it can only occur for "fresh" tasks which don't have a stack allocation yet; so I assume we could just do `if (task->copy_stack ==1) return;` or so?

Comment (rbehrends) [reply src/julia_gc.c:689]: Hmm, I'll go and have a look at that.

Comment (rbehrends) [reply src/julia_gc.c:689]: So, `task->copy_stack == 1` implies `task->bufsz == 0`. This means that the logic is accidentally correct, but the comment would have to change. But I think that making control flow more explicit by aborting stack scanning for `task->copy_stack == 1` is much clearer, so I'll go with that.

File: src/julia_gc.c [hunk 15]
```diff
@@ -485,13 +518,16 @@ static void SafeScanTaskStack(PtrArray * stack, void * start, void * end)
...
-        // guard pages; accessing these results in segmentation
-        // faults. Julia catches those segmentation faults and
-        // longjmps to JuliaTLS->safe_restore; we use this
-        // mechamism to abort stack scanning when a protected
-        // page is hit. For this to work, we must scan the stack
-        // from top to bottom, so we see any guard pages last.
+        // The start of the stack buffer may be protected with guard
+        // pages; accessing these results in segmentation faults.
+        // Julia catches those segmentation faults and longjmps to
+        // JuliaTLS->safe_restore; we use this mechamism to abort stack
```
Comment (fingolfin) [inline src/julia_gc.c:524]:
```suggestion
        // JuliaTLS->safe_restore; we use this mechanism to abort stack
```

**Comment by fingolfin** [comment]
Done in 74cfb2a8f66a31370aaa7cfd9bd05dbc563ccea6
